File: Java/Dagger/atm/src/main/java/ro/razvan/java/dagger/atm/BigDecimalCommand.java
package ro.razvan.java.dagger.atm;

import java.math.BigDecimal;

public interface BigDecimalCommand extends SingleArgCommand {

    @Override
    default Result handleArg(String arg) {
        try {
            return handleAmount(new BigDecimal(arg));
        } catch (NumberFormatException exc) {
            return Result.invalid();
        }
    }

    Result handleAmount(BigDecimal amount);
}

File: Kotlin/Dagger/atm/src/main/kotlin/ro/razvan/kotlin/dagger/atm/BigDecimalCommand.kt
package ro.razvan.kotlin.dagger.atm

import ro.razvan.kotlin.dagger.atm.Command.Result
import java.math.BigDecimal

interface BigDecimalCommand : SingleArgCommand {

    override fun handleArg(arg: String): Result {
        arg.toBigDecimalOrNull()?.let(::handleAmount)
        return Result.handled()
    }

    fun handleAmount(amount: BigDecimal)

}